File: Mac/ServikoMobile/ServikoMobile/PriceTree.swift
//
//  PriceTree.swift
//  ServikoMobile
//
//  Created by ert on 13.10.2020.
//

import Foundation

class TreeElement: Comparable {

    static func < (lhs: TreeElement, rhs: TreeElement) -> Bool {
        return lhs.item.name < rhs.item.name
    }
    
    static func == (lhs: TreeElement, rhs: TreeElement) -> Bool {
        return lhs.item.name == rhs.item.name
    }
    
    var item : Price

    init(src: Price) { self.item = src }
}

class Folder : TreeElement {
    var childs = [Folder]()
    var items = [TreeElement]()
    
    init() {
        super.init(src: Price())
    }
    
    override init(src: Price) {
        super.init(src:src)
    }

    var count : Int {  get { return childs.count + items.count } }

    func setSrc(src: Price) { self.item = src; }

    func add(folder: Folder) { childs.append(folder) }
    
    func add(price: Price) {
        let te = TreeElement(src:price);
        items.append(te);
    }

    subscript(index: Int) -> TreeElement {
        return index < childs.count ? childs[index] : items[index - childs.count]
    }
    

    func sort() {
        childs.sort()
        items.sort()
    }

    func findItem( dest:inout [TreeElement], pattern:String) {
        for el in items {
            if el.item.name.range(of: pattern, options: [.caseInsensitive, .regularExpression]) != nil {
                dest.append(el)
            }
        }
        
        for f in childs {
            f.findItem(dest:&dest, pattern:pattern)
        }
    }
   
    func findFolder(id:String) -> Folder? {
        if item.id == id {
            return self
        }
        for f in childs {
            if let fndF = f.findFolder(id:id) {
                return fndF
            }
        }
        
        return nil
    }

    func findItem(id:String) -> Price? {
        for i in items {
            if i.item.id == id {
                return i.item
            }
        }
        
        for f in childs {
            if let pi = f.findItem(id: id) {
                return pi
            }
        }
        
        return nil
    }
}

class PriceTree {
    let root = Folder()

    func find(id: String) -> Price? {
        return root.findItem(id:id);
    }

    func findFolder(id: String) -> Folder? {
        return root.findFolder(id: id);
    }
    
    static func make(data : [Price]) -> PriceTree {
        let ret = PriceTree()
        var folders = [String:Folder]()
        folders[""] =  ret.getRoot()

        var mapPrice = [String:Price]()

        for i in data {
            mapPrice[i.id] = i

            var parent = folders[i.parent]
            if(parent == nil) {
                parent = Folder()
                folders[i.parent] = parent
            }

            if(i.isFolder) {
                var dest = folders[i.id]
                if(dest == nil) {
                    dest = Folder(src:i);
                    folders[i.id] = dest
                } else {
                    dest!.setSrc(src:i);
                }
                parent!.add(folder: dest!)
            } else {
                parent!.add(price:i)
            }
        }
        for (_, f) in folders  {
            f.sort();
        }
        return ret;
    }

    var count: Int { get { return root.count } }
    
    func getRoot() -> Folder { return root }
}

File: Napoleon.ce/Core/Napoleon.AS/serviko.sales/app/src/main/java/com/serviko/dataobjects/priceTree/PriceTree.java
package com.serviko.dataobjects.priceTree;

import com.serviko.dataobjects.Order;
import com.serviko.dataobjects.OrderItem;
import com.serviko.dataobjects.Price;
import com.serviko.dataobjects.PriceCategory;
import com.serviko.dataobjects.PriceCategoryFolder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PriceTree {
    Folder root = new Folder();

    public Price find(String id) {
        return root.findItem(id);
    }

    public Folder findFolder(String id) {
        return root.findFolder(id);
    }

    public static PriceTree make(List<Price> data, Set<String> manufacturer) {
        PriceTree ret = new PriceTree();
        Map<String, Folder> folders = new HashMap<>();
        folders.put("", ret.root);

        for(Price p : data) {
            if(p.isFolder) continue;

            manufacturer.add(p.suplyer);

            Map<Integer, Folder> cf = new TreeMap<>();

            List<PriceCategoryFolder> pcf = new ArrayList<>();
            for (PriceCategory pc : p.categories) {
                PriceCategoryFolder ce = new PriceCategoryFolder(pc);
                if (ce.isEmpty()) {
                    ce.name = "<>";
                    ce.code = ce.name;
//                    continue;
                }
                pcf.add(ce);
            }
            while(pcf.size() < 2) {
                PriceCategoryFolder ce = new PriceCategoryFolder(new PriceCategory());
                ce.name = "<?>";
                ce.level = pcf.size() + 1;
                ce.code = ce.name + "/" + Integer.toString(ce.level);
                pcf.add(ce);
            }

            Collections.sort(pcf);
            String folderCode = "";
            for(PriceCategoryFolder ce : pcf) {
                ce.code += '\t' + Integer.toString(ce.level);
                if(folderCode.length() > 0)
                    ce.code = folderCode + '\t' + ce.code;
                Folder f = folders.get(ce.code);
                if(f == null) {
                    f = new Folder(ce);
                    folders.put(ce.code, f);
                }
                cf.put(ce.level, f);
                folderCode = ce.code;
            }
            Folder curf = ret.root;
            for(Map.Entry<Integer,Folder> kv : cf.entrySet()) {
                curf = curf.add(kv.getValue());
            }
            curf.add(p);
        }

        return ret;
    }

    public static PriceTree make1(List<Price> data) {
        PriceTree ret = new PriceTree();
        Map<String, Folder> folders = new HashMap<>();
        folders.put("", ret.root);

        Map<String, Price> mapPrice = new HashMap<>();

        for(Price i : data) {
            mapPrice.put(i.id, i);

            Folder parent = folders.get(i.parent);
            if(parent == null) {
                parent = new Folder();
                folders.put(i.parent, parent);
            }

            if(i.isFolder) {
                Folder dest = folders.get(i.id);
                if(dest == null) {
                    dest = new Folder(i);
                    folders.put(i.id, dest);
                } else {
                    dest.setSrc(i);
                }
                parent.add(dest);
            } else {
                parent.add(i);
            }
        }
        for(Folder f : folders.values()) {
            f.sort();
        }
        return ret;
    }

    public PriceTree() {}

    public int size() { return root.size(); }
    public Folder root() { return root; }
}
